### DISCUSSION TITLE
More settings/improvements when scaling is used

### DISCUSSION THREAD

**Discussion by Moltanicaa** [ppy/osu#17611] [Ideas]
When scaling is active, the empty area is filled with the background of the map. It completely ignores the set dimming settings and does not allow you to make it completely black. This increases the amount of resources consumed.

This seems to be a duplicate of this https://github.com/ppy/osu/discussions/13687 thread, but I'm talking about something a little different. In this regard, there is a small wish - is it possible to make a frame around the playing field in order to visually see the boundaries of the playing field?
![161324276-bcee653a-b05d-4c67-9aed-6f098d8b51ee](https://user-images.githubusercontent.com/59089206/161375119-3b36a313-6387-40a2-a4e1-36f8b2d45d63.jpg)

**Comment by smoogipoo** [comment]
I'm not sure what the point of this discussion is. Is it:
1) You want to have more settings related to the fullscreen background (such as dimming, as you mentioned), and a border around the inner contents, or...
2) You're bringing up the performance concerns?

In regards to the performance concerns, it should be mostly irrelevant since it's 2x draw and _should_ be covered by the front-to-back render pass. However I found two points of concern that prevent this:
1. Textured sprites no longer contribute to the FTB pass because we've disabled the code that determines if the texture is opaque: https://github.com/ppy/osu-framework/blob/d5b072fc890ffcfccc6812170307d69c0071fcc0/osu.Framework/Graphics/OpenGL/Textures/TextureGL.cs#L154-L179
2. `BufferedContainer` currently doesn't know whether its contents are fully opaque or not, and so also doesn't contribute to the FTB pass.

**Reply by Moltanicaa** [reply]
I'm only talking about dimming settings and a line around the screen. About performance it is said according to the principle of "specking of witch". If you found an additional problem yourself, then okay, but this was not the main goal.

**Reply by smoogipoo** [reply]
Thanks for the clarification.

**Reply by cdwcgt** [reply]
I made a simple implementation that can change the dim of the background

https://user-images.githubusercontent.com/34775378/184644786-dfff4a0a-aa69-490d-8500-3ff35391d238.mp4

**Reply by cdwcgt** [reply]
[Here is my branch](https://github.com/cdwcgt/osu/tree/scale-background-dim)
I asked at discord for opinion and seen like may not be particularly good

I'm here hoping for a straight answer because I don't want to open a pr before confirm

**Reply by peppy** [reply]
As per my reply in discord, I think a toggle would work better. I don't think anyone would want to make the background brighter, and the headroom for making darker is so small I think you should hide it at that point.

**Reply by cdwcgt** [reply]
I think an adjustable slider would give the player more options
just like beatmap background dim
And also in line with the design of scaling Settings

**Reply by peppy** [reply]
More is not always better.

Please re-read my message.

**Reply by cdwcgt** [reply]
So either I prove that this feature is liked by a lot of people, or change it to a toggle right?
To be honest, instead of having it with a default value of 90%, removing it would also avoid rendering performance issues.
90% of the background dim is hard to notice that there is this feature even if the background is bright

issues about focus
> https://discord.com/channels/188630481301012481/188630652340404224/1008735469690945619

set beatmap background dim to 100%(or other make it black enough) but the semi-dark border allows players to focus more on the field
in this discussion thread [Moltanicaa](https://github.com/Moltanicaa) have a request
> is it possible to make a frame around the playing field in order to visually see the boundaries of the playing field?

![image](https://user-images.githubusercontent.com/34775378/184793582-0a4e9e53-4f0d-4d23-a078-9ab8f052796b.png)
(50dim setting)
I think this solves the problem nicely

and for performance issues, https://github.com/cdwcgt/osu/commit/6113c75469f76211d4b81edd070ab08617f5aac5
I make `requiresBackgroundVisible` false if dim is set to 100%
This should make lazer stop rendering the background like `ScalingMode.Gameplay`

If it had to be a toggle anyway, then we may need to give up 90% because it is meaningless because player who want to adjust it need more lower value, or give up 0% but it's also meaningless because they prefer to close it totally or not 90% dim just like player need 100% dim for beatmap dim or not 99%

I thought of a way to compromise
we can change the minimum value of the slider to 30% (old stable defaults for beatmap dim) or not 0%

**Reply by peppy** [reply]
The request you mention is about the *playfield*, not the game window. This does not solve that issue, and we have a border already implemented to show the edge of the play area.

**Reply by cdwcgt** [reply]
ya I know there is a settings of *playfield border style* but according to the screenshot author may misunderstanding **overlay** as **playfield**
He is referring to the range of mouse movement during gaming
In the stable version, people who don't touch the editor are highly likely to misunderstand it

And I also tried it myself for a long time with 100% beatmap background dim and 50% scale background dim. This doesn't cause distraction for the player, on the contrary it may relieve tension

**Reply by peppy** [reply]
Sure. Will wait for more opinions.

**Reply by Moltanicaa** [reply]
As far as I understand from your discussion, there is some misunderstanding regarding the request. I am writing specifically about displaying the border of the active game ZONE, not the playing field (which, by the way, works very poorly, since it does not even closely display the real borders of the field)
The game zone includes all game elements + the playing field. With 100% dimming of the entire screen (including the area behind the game ZONE) and active screen scaling, I have no way to see the actual boundaries of the game ZONE (NOT THE FIELD). When playing, it is very important to visually understand how big or small the playing area is at the moment.
![Безымянный](https://user-images.githubusercontent.com/59089206/184892364-46feca0b-932d-4f09-882b-674041baa66c.png)

Actually the idea of being able to reduce the dimming and add a blur behind the background (the one behind the game ZONE) is a good idea, especially if you let it be customizable. But it will consume more resources.

**Reply by Moltanicaa** [reply]
Since we are talking about the border of the playing field here, can I make a separate discussion of the problem of its current implementation? The edge of the field is in the center of the closest point to the edge where a note can be placed, which is why it is very common for a note/slider to be outside the edge of this field.


UPD. https://github.com/ppy/osu/discussions/13167

**Reply by cdwcgt** [reply]
What you may need is a specific zone of objects placed on the **edge**
this is related to cs, is cs higher the object will more smaller. The actual playing field will also be smaller

It doesn't seem to be in the same direction as our discussion, maybe I need to open another discussion

**Reply by bdach** [reply]
@Moltanicaa please do not derail discussion with unrelated issues. open a separate discussion (although I'm pretty sure one covering the issue you mention will already exist at this point)

**Reply by cdwcgt** [reply]
> Sure. Will wait for more opinions.

I've been thinking about this for the past few days. We might need a kind of poll? 
I'm not sure because I personally want to initiate this kind of investigation but I don't have that much influence.
Then I thought about making the osu official account(or peppy's twitter?) but maybe it's not necessary to do so
But it feels unrealistic to wait here for other players to respond. So I'm at a loss right now

Much of it depends on ppy's team, But I really hope to get an accurate answer as soon as possible.

**Comment by cdwcgt** [comment]
https://github.com/ppy/osu/pull/23038
